sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java

```diff
@@ -203,17 +201,73 @@ private void processWhere() {
                 qs.where(qs.where().add(queryForRelation));
             }
         }
+
+        tryToConvertImplicitJoinConditionsToJoinPairs(joinPairs, splitQueries);
         if (splitQueries.isEmpty()) {
             querySpec.where(WhereClause.MATCH_ALL);
         } else {
             Symbol newQuery = AndOperator.join(splitQueries.values());
             JoinConditionValidator.validate(newQuery);
             querySpec.where(new WhereClause(newQuery));
         }
+
+        processJoinPairs();
+    }
+
+    private void processJoinPairs() {
+        for (JoinPair joinPair : joinPairs) {
+            if (joinPair.condition() != null) {
+                JoinConditionValidator.validate(joinPair.condition());
+                joinConditions.add(joinPair.condition());
+                relationPartOfJoinConditions.add(joinPair.left());
+                relationPartOfJoinConditions.add(joinPair.right());
+            }
+        }
+    }
+
+    @VisibleForTesting
+    static void tryToConvertImplicitJoinConditionsToJoinPairs(List<JoinPair> joinPairs,
```

Comment:
Similar here. Would be nice if we could have a `splitQueries -> joinPairs` signature. Without side affects or mutations.

Reply:
You're right, I thought about that but I didn't do in the first place because the splitQueries need to be mutated and I also chose (probaly wrong) to mutate the joinPairs too